sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java

```diff
@@ -0,0 +1,595 @@
+              }
+              return new Object[]{
+                  val.getKey().getIdentifier(),
+                  val.getKey().getDataSource(),
+                  val.getKey().getInterval().getStart(),
+                  val.getKey().getInterval().getEnd(),
+                  val.getKey().getSize(),
+                  val.getKey().getVersion(),
+                  val.getKey().getShardSpec().getPartitionNum(),
+                  val.getValue().getNumReplicas(),
+                  val.getValue().getNumRows(),
+                  val.getValue().isPublished(),
+                  val.getValue().isAvailable(),
+                  val.getValue().isRealtime(),
+                  jsonMapper.writeValueAsString(val.getKey())
+              };
+            }
+            catch (JsonProcessingException e) {
+              log.error(e, "Error getting segment payload for segment %s", val.getKey().getIdentifier());
+              throw new RuntimeException(e);
+            }
+          });
+
+      final FluentIterable<Object[]> publishedSegments = FluentIterable
+          .from(() -> metadataSegments)
+          .transform(val -> {
+            try {
+              if (availableSegmentIds.contains(val.getIdentifier())) {
+                return null;
+              }
+              return new Object[]{
+                  val.getIdentifier(),
+                  val.getDataSource(),
+                  val.getInterval().getStart(),
+                  val.getInterval().getEnd(),
+                  val.getSize(),
+                  val.getVersion(),
+                  val.getShardSpec().getPartitionNum(),
+                  0l,
+                  -1l,
+                  1l,
+                  0l,
+                  0l,
+                  jsonMapper.writeValueAsString(val)
+              };
+            }
+            catch (JsonProcessingException e) {
+              log.error(e, "Error getting segment payload for segment %s", val.getIdentifier());
+              throw new RuntimeException(e);
+            }
+          });
+
+      Iterable<Object[]> allSegments = Iterables.unmodifiableIterable(
+          Iterables.concat(availableSegments, publishedSegments));
+
+      return Linq4j.asEnumerable(allSegments).where(t -> t != null);
+
+    }
+  }
+
+  // Note that coordinator must be up to get segments
+  private static JsonParserIterator<DataSegment> getMetadataSegments(
+      DruidLeaderClient coordinatorClient,
+      ObjectMapper jsonMapper,
+      BytesAccumulatingResponseHandler responseHandler
+  )
+  {
+
+    Request request;
+    try {
+      request = coordinatorClient.makeRequest(
+          HttpMethod.GET,
+          StringUtils.format("/druid/coordinator/v1/metadata/segments")
+      );
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    ListenableFuture<InputStream> future = coordinatorClient.goStream(
+        request,
+        responseHandler
+    );
+    try {
+      future.get();
```

Comment:
This would block until it gets all results. Is this intentional?

Reply:
Yes, this is intentional. Checked the query performance on the test cluster and it was okay.

Reply:
Ah, sorry. My question is why `future` needs to get here. `get` is called anyway in `JsonParserIterator`.

The query performance should be good because `/druid/coordinator/v1/metadata/segments` returns the `StreamingOutput`.

Reply:
That makes sense, removed the `future.get` from here and `JsonParserIterator` will do the `get`